include/openthread/dns_client.h

```diff
@@ -55,9 +55,64 @@ extern "C" {
  *
  */
 
-#define OT_DNS_DEFAULT_SERVER_PORT 53 ///< The default DNS Server port.
+/**
+ * This enumeration type represents the "Recursion Desired" (RD) flag in an `otDnsQueryConfig`.
+ *
+ */
+typedef enum
+{
+    OT_DNS_FLAG_UNSPECIFIED       = 0, ///< Indicates the flag is not specified.
+    OT_DNS_FLAG_RECURSION_DESIRED = 1, ///< Indicates DNS name server can resolve the query recursively.
+    OT_DNS_FLAG_NO_RECURSION      = 2, ///< Indicates DNS name server can not resolve the query recursively.
+} otDnsRecursionFlag;
+
+/**
+ * This structure represents a DNS query configuration.
+ *
+ * Any of the fields in this structure can be set to zero to indicate that it is not specified.
+ *
+ */
+typedef struct otDnsQueryConfig
+{
+    otSockAddr         mServerSockAddr;  ///< Server address (IPv6 address/port). All zero or zero port for unspecified.
+    uint32_t           mResponseTimeout; ///< Wait time (in msec) to rx response. Zero indicates unspecified value.
+    uint8_t            mMaxTxAttempts;   ///< Maximum tx attempts before reporting failure. Zero for unspecified value.
+    otDnsRecursionFlag mRecursionFlag;   ///< Indicates whether the server can resolve the query recursively or not.
+} otDnsQueryConfig;
+
+/**
+ * This function gets the current default query config used by DNS client.
+ *
+ * When OpenThread stack starts, the default DNS query config is determined from a set of OT config options such as
+ * `OPENTHREAD_CONFIG_DNS_CLIENT_DEFAULT_SERVER_IP6_ADDRESS`, `_DEFAULT_SERVER_PORT`, `_DEFAULT_RESPONSE_TIMEOUT`, etc.
+ * (please see `config/dns_clinet.h` for all related config options).
+ *
+ * @param[in]  aInstance        A pointer to an OpenThread instance.
+ *
+ * @returns A pointer to the current default config being used by DNS client.
+ *
+ */
+const otDnsQueryConfig *otDnsClientGetDefaultConfig(otInstance *aInstance);
```

Comment:
```suggestion
const otDnsQueryConfig *otDnsClientGetConfig(otInstance *aInstance);
```

There seems to be 3 configs:
- The starting default config when OT starts.
- The default config that is configurable 
- The config for each query

Would it be more clear if we change it to only 2 configs ? 

- The starting default config when OT starts.
- The current config that is configurable

Then each browse or resolve query would just use the current config without specifying a config.

Reply:
Note that the config parameter in each query is optional and caller can choose not to use/provide one.

This model gives the user flexibility to set a default server/config  (e.g. the one used for all DNS-SD queries) so service discovery can be done easier, but still be able to send query to other servers (specify config per query) without changing the default config.

> Remove this argument since the caller can already do everything it want with otDnsClientSetDefaultConfig (which I suggested to rename to otDnsClientSetConfig) ?
> I think we should make the API as simple as possible, unless the benefits are clear.

There is a reason and benefit for this model:
- User (next layer code) may not want to use same server for all query types:
   - e.g., address query may use a differernt name server, compared to the server used for DNS-SD queries (another use-case is nat64 which may require its own address queries).
- If we require all queries to use the default config, then the user need to call multiple APIs whenever she wants to send a query with a different config:
  -  First change the default config, start the query, change it back to original 
  - or user needs to always call two APIs: set the default, then call the query (which defeats the purpose of having a default config).

Reply:
Agree that the current model has better flexibility and so, no strong preference.

> e.g., address query may use a differernt name server, compared to the server used for DNS-SD queries (another use-case is nat64 which may require its own address queries).

The DNS-SD server could recursively query other name servers so that the DNS client don't need to use different DNS server for different purpuses.

> If we require all queries to use the default config, then the user need to call multiple APIs whenever she wants to send a query with a different config:

To reduce code size and complexity, I tend to prefer the strategy of not implementing features that are unclear if it's going to be used or how it's going to be used, until real use cases appear. 

Again, no strong preference. So, I am resolving this conversation.